File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/other/GoogleCaptchaFragment.java
package forpdateam.ru.forpda.ui.fragments.other;

import android.app.Activity;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.regex.Pattern;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.api.NetworkRequest;
import forpdateam.ru.forpda.api.NetworkResponse;
import forpdateam.ru.forpda.client.Client;
import forpdateam.ru.forpda.common.webview.CustomWebViewClient;
import forpdateam.ru.forpda.ui.activities.MainActivity;
import forpdateam.ru.forpda.ui.fragments.TabFragment;
import forpdateam.ru.forpda.ui.views.ExtendedWebView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by radiationx on 09.11.17.
 */

public class GoogleCaptchaFragment extends TabFragment {
    private ExtendedWebView webView;
    private String content = "";

    public GoogleCaptchaFragment() {
        configuration.setDefaultTitle("Проверка");
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            content = getArguments().getString("content", "1");
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        webView = getMainActivity().getWebViewsProvider().pull(getContext());
        attachWebView(webView);
        fragmentContent.addView(webView);
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewsReady();
        setSubtitle("Это из-за VPN/Proxy и т.д.");
        webView.setWebViewClient(new CaptchaWebViewClient());
    }

    class CaptchaWebViewClient extends CustomWebViewClient {
        @Override
        public boolean handleUri(Uri uri) {
            Log.e("SUKA", uri.toString());
            if (Pattern.compile("https://4pda.ru/cdn-cgi/l/chk_captcha").matcher(uri.toString()).find()) {
                NetworkRequest nr = new NetworkRequest.Builder().url(uri.toString()).withoutBody().build();
                Observable.fromCallable(() -> Client.get().request(nr))
                        .onErrorReturn(throwable -> new NetworkResponse(null))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(GoogleCaptchaFragment.this::onResponse);
            }
            return true;
        }
    }

    @Override
    public boolean loadData() {
        super.loadData();
        webView.loadDataWithBaseURL("https://4pda.ru/forum/", content, "text/html", "utf-8", null);
        return true;
    }

    private void onResponse(NetworkResponse response) {
        Toast.makeText(App.getContext(), "Приложение будет перезапущено", Toast.LENGTH_SHORT).show();
        new Handler().postDelayed(() -> {
            Activity activity = App.getActivity();
            if (activity == null) {
                Toast.makeText(App.getContext(), "Перезапустите приложение", Toast.LENGTH_SHORT).show();
            }
            MainActivity.restartApplication(activity);
        }, 1000);
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/other/GoogleCaptchaFragment.kt
package forpdateam.ru.forpda.ui.fragments.other

import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.View
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Toast
import androidx.core.net.toUri
import androidx.lifecycle.lifecycleScope
import forpdateam.ru.forpda.common.Utils
import forpdateam.ru.forpda.common.webview.DialogsHelper
import forpdateam.ru.forpda.model.data.remote.WebClient
import forpdateam.ru.forpda.model.data.remote.api.NetworkRequest
import forpdateam.ru.forpda.model.data.remote.api.common.CaptchaParser
import forpdateam.ru.forpda.presentation.LinkHandler
import forpdateam.ru.forpda.presentation.SystemLinkHandler
import forpdateam.ru.forpda.presentation.TabRouter
import forpdateam.ru.forpda.ui.activities.MainActivity
import forpdateam.ru.forpda.ui.fragments.TabFragment
import forpdateam.ru.forpda.ui.views.ExtendedWebView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import ru.radiationx.quill.inject

/**
 * Created by radiationx on 09.11.17.
 */

class GoogleCaptchaFragment : TabFragment() {

    companion object {
        fun newInstance() = GoogleCaptchaFragment()
    }

    private lateinit var webView: ExtendedWebView
    private var content = ""

    private val utils by inject<Utils>()
    private val linkHandler by inject<LinkHandler>()
    private val systemLinkHandler by inject<SystemLinkHandler>()
    private val router by inject<TabRouter>()
    private val webClient by inject<WebClient>()
    private val captchaParser by inject<CaptchaParser>()

    init {
        configuration.defaultTitle = "Проверка"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.apply {
            content = getString("content", "1")
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        webView = ExtendedWebView(requireContext())
        webView.setDialogsHelper(
            DialogsHelper(
                context = webView.context,
                linkHandler = linkHandler,
                systemLinkHandler = systemLinkHandler,
                utils = utils,
                router = router
            )
        )
        attachWebView(webView)
        fragmentContent.addView(webView)

        setSubtitle("Это из-за VPN/Proxy и т.д.")
        webView.webViewClient = CaptchaWebViewClient()
        webView.loadDataWithBaseURL("https://4pda.to/forum/", content, "text/html", "utf-8", null)
    }

    internal inner class CaptchaWebViewClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            return handleUri(request.url)
        }

        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            return handleUri(url.toUri())
        }

        private fun handleUri(uri: Uri): Boolean {
            Log.e("SUKA", uri.toString())
            if (captchaParser.checkRedirect(uri.toString())) {
                runBlocking {
                    runCatching {
                        val nr = NetworkRequest.Builder().url(uri.toString()).withoutBody().build()
                        webClient.request(nr)
                    }
                    withContext(Dispatchers.Main) {
                        this@GoogleCaptchaFragment.onResponse()
                    }
                }
            }
            return true
        }
    }

    private fun onResponse() {
        viewLifecycleOwner.lifecycleScope.launch {
            val activity = requireActivity()
            Toast.makeText(activity, "Приложение будет перезапущено", Toast.LENGTH_SHORT).show()
            delay(1000)
            MainActivity.restartApplication(activity)
        }
    }
}
